### PR TITLE
[WIP] Add cudnn fused multi-head attention for JAX

### PR DESCRIPTION
**Pull request by zlsh80826**
1. Move `scale_factor` to core attention to align fused multi-head attention implementation [transformer.py, module.py]
2. Add cudnn-frontend submodule into `3rdparty/cudnn-frontend`
3. Add cudnn-frontend fused multi-head attention with both self attention and cross attention
4. Fused multi-head attention will be auto enabled if the network satisify the following rules
    - not decode
    - not transpose_batch_sequence
    - fuse_qkv
    - dropout_rate = 0 (dropout can be fused into FMHA, but we lack model convergence test. Will add it in the future)
    - dtype = bfloat16 or float16
    - q_seqlen and kv_seqlen = 128 or 256 or 384 or 512
5. MHA supports DP and TP sharding

### CODE REVIEW

File: transformer_engine/common/include/transformer_engine/fmha.h
Comment (ptrendx) [inline transformer_engine/common/include/transformer_engine/fmha.h:1]: The proposed API is completely different than the rest of TE APIs and is not acceptable.

@cyanguwa is working on a better abstraction for multiple different fused MHA algorithms, please coordinate with her to plug under that API.

**Review by ptrendx — CHANGES_REQUESTED** [review]

File: transformer_engine/common/include/transformer_engine/fmha.h
Comment (zlsh80826) [reply transformer_engine/common/include/transformer_engine/fmha.h:1]: OK, I will work with @cyanguwa to see how to integrate the FMHA

File: transformer_engine/jax/fmha.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+from .cpp_extensions import self_fmha_fwd, self_fmha_bwd
+from .sharding import get_fmha_sharding_meta
+from .sharding import ShardingType
+from .sharding import xmap_runner
+
+jax.config.update('experimental_xmap_spmd_lowering', True)
+jax.config.update('experimental_xmap_spmd_lowering_manual', True)
+
+
+def self_fmha(qkv: jnp.ndarray,
```
Comment (ptrendx) [inline transformer_engine/jax/fmha.py:20]: Is this the public API that people should use? If it is, that is not great:
 - the name is not descriptive - what is "fmha"?
 - what it implements is not "Multihead attention" but rather "Dot Product Attention" from the "Attention is all you need paper", so should not really use mha in the name
 - docstrings are missing
 - it will not extend to the non-zero dropout case since there you need offset too. Even if we do not implement that case right away, we should nto expose the API that will prevent us from doing so in the future.
 - what is "scaling factor"?
 - why isn't it a module like the other high-level APIs?

Comment (zlsh80826) [reply transformer_engine/jax/fmha.py:20]: - The public API that people should use is [MultiHeadAttention](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/transformer.py#L188)([TE API doc](https://nvidia.github.io/TransformerEngine/api/jax.html#transformer_engine.jax.MultiHeadAttention)), the `self_fmha` and `cross_fmha` are custom_calls used internally for JAX-TE instead of public users.
- The `"fmha"` naming is inherited by apex, I can rename them to `self(cross)_fused_dot_product_attention` if they are preferred in TE.
- As described above, the custom_calls are not intented to public users. The doc strings for MultiHeadAttetion are written on [MultiHeadAttention](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/transformer.py#L188)
- The [original cuDNN sample](https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/fused_mha_sample.cpp#L658) has missed the offset and used a CPU-side seed. But yes, we should use both seed and offset and keep them as the device pointers.
- Is `scaling_factor` ambiguous here? From the "Attention Is All You Need paper", the scaling factor is `rsqrt(head_dim)` that scaled to `dot_product(Q, tranpose(K))`
- [MultiHeadAttention](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/transformer.py#L188)([TE API doc](https://nvidia.github.io/TransformerEngine/api/jax.html#transformer_engine.jax.MultiHeadAttention)) is implemented as a module. self_fmha is a custom_call wrapper, it is the same level as [fp8_dot](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/dot.py#L23)

I will add some commits for fmha renaming and seed/offset changes. About the naming, which is preferred in TE? `self(cross)_fused_dot_product_attention`, `self(cross)_fused_attention` or other?

Comment (ptrendx) [reply transformer_engine/jax/fmha.py:20]: Oh boy... This means we have a divergence between pyTorch and JAX APIs. pyTorch doesn't expose the MultiHeadAttention API and now this kind of forces us to do that. @ksivaman for visibility. Still, for consistency I would really like for JAX to expose DotProductAttention to be in line with pyTorch (and then of course the MultiHeadAttention API can use that similarly to how it uses the other modules).

If those are not public APIs then most of my comments are not applicable. I would probably settle for `self/cross_fused_attention`.

About scaling factor - sorry, my bad, I forgot about this scaling and thought about this qk layer scaling that megatron uses. It's fine as is :+1:.

Comment (zlsh80826) [reply transformer_engine/jax/fmha.py:20]: Yes, I will modularize `DotProductAttention` for JAX in the coming days before integrating the fused attentions.

**Comment by timmoon10** [comment]
PyTorch support for fused attention is added in https://github.com/NVIDIA/TransformerEngine/pull/155.

**Comment by zlsh80826** [comment]
Close this PR and I will open a new one for the fp16/bf16 fused attention (max_seq <= 512)
